src/rmp/src/zero_slack_strategy.cpp

```diff
@@ -0,0 +1,98 @@
+// Copyright 2025 Google LLC
+//
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file or at
+// https://developers.google.com/open-source/licenses/bsd
+
+#include "zero_slack_strategy.h"
+
+#include <iostream>
+
+#include "abc_library_factory.h"
+#include "db_sta/dbNetwork.hh"
+#include "db_sta/dbSta.hh"
+#include "delay_optimization_strategy.h"
+#include "logic_cut.h"
+#include "logic_extractor.h"
+#include "sta/Graph.hh"
+#include "sta/PortDirection.hh"
+#include "sta/Units.hh"
+#include "sta/VerilogWriter.hh"
+#include "unique_name.h"
+#include "utl/Logger.h"
+#include "utl/deleter.h"
+
+namespace rmp {
+
+std::vector<sta::Vertex*> GetNegativeVertices(sta::dbSta* sta)
+{
+  std::vector<sta::Vertex*> result;
+
+  sta::dbNetwork* network = sta->getDbNetwork();
+  for (sta::Vertex* vertex : *sta->endpoints()) {
+    sta::PortDirection* direction = network->direction(vertex->pin());
+    if (!direction->isInput()) {
+      continue;
+    }
+
+    sta::Slack slack = sta->vertexSlack(vertex, sta::MinMax::max());
```

Comment:
Use const where possible throughout

Reply:
I'm probably misunderstanding something, but it looks like you are collecting _input_ vertices with negative slack? Is that what you want?

Reply:
The endpoints are generally the D and Q ports of flip flops. We're trying to grab a cone of logic which we want to start at D ports. So in this case we do want the inputs ports of flops.

Reply:
What about cones from primary outputs?

Reply:
I think there's still an input pin there if I'm not mistaken.

Reply:
What would it be an input to?  I would verify that.

Reply:
Ah, yes, of course. It's just that I had the default Yosys behavior in mind. They do not send any sequentials to ABC.

Reply:
Done